File: services/matchmaking/test/integration_test.py
#!/usr/bin/env python3
"""
Integration test for matchmaking service.
Tests the full flow: register users -> join queue -> verify match.
"""

import requests
import time
import sys
import random
import string

AUTH_SERVICE_URL = "http://localhost:8000"
MATCHMAKING_SERVICE_URL = "http://localhost:8001"


def generate_random_username():
    """Generate a random username to avoid conflicts."""
    return f"testuser_{''.join(random.choices(string.ascii_lowercase + string.digits, k=8))}"


def register_user(username, password):
    """Register a new user via auth service."""
    url = f"{AUTH_SERVICE_URL}/auth/register"
    payload = {
        "username": username,
        "password": password
    }

    print(f"📝 Registering user: {username}")
    response = requests.post(url, json=payload)

    if response.status_code not in [200, 201]:
        print(f"❌ Failed to register {username}: {response.status_code}")
        print(f"   Response: {response.text}")
        sys.exit(1)

    data = response.json()
    print(f"✅ Registered {username} with user_id: {data['user_id']}")
    return data


def login_user(username, password):
    """Login a user and get auth token."""
    url = f"{AUTH_SERVICE_URL}/auth/login"
    payload = {
        "username": username,
        "password": password
    }

    print(f"🔐 Logging in user: {username}")
    response = requests.post(url, json=payload)

    if response.status_code != 200:
        print(f"❌ Failed to login {username}: {response.status_code}")
        print(f"   Response: {response.text}")
        sys.exit(1)

    data = response.json()
    print(f"✅ Logged in {username}, token: {data['token'][:16]}...")
    return data


def join_queue(token, username):
    """Add a user to the matchmaking queue."""
    url = f"{MATCHMAKING_SERVICE_URL}/matchmaking/join"
    headers = {"Authorization": f"Bearer {token}"}

    print(f"🎮 {username} joining queue...")
    response = requests.post(url, json={}, headers=headers)

    if response.status_code != 200:
        print(f"❌ Failed to join queue for {username}: {response.status_code}")
        print(f"   Response: {response.text}")
        sys.exit(1)

    data = response.json()
    print(f"✅ {username} joined queue (MMR: {data.get('mmr', 'N/A')})")
    return data


def check_match(token, username):
    """Check if a user has been matched."""
    url = f"{MATCHMAKING_SERVICE_URL}/matchmaking/match"
    headers = {"Authorization": f"Bearer {token}"}

    response = requests.get(url, headers=headers)

    if response.status_code != 200:
        print(f"❌ Failed to check match for {username}: {response.status_code}")
        print(f"   Response: {response.text}")
        sys.exit(1)

    return response.json()


def accept_match(session_id, token, username):
    """Accept a match."""
    url = f"{MATCHMAKING_SERVICE_URL}/matchmaking/accept"
    headers = {"Authorization": f"Bearer {token}"}
    payload = {"session_id": session_id}

    print(f"✅ {username} accepting match {session_id[:8]}...")
    response = requests.post(url, json=payload, headers=headers)

    if response.status_code != 200:
        print(f"❌ Failed to accept match for {username}: {response.status_code}")
        print(f"   Response: {response.text}")
        sys.exit(1)

    data = response.json()
    print(f"   Response: {data.get('message', 'N/A')}")
    return data


def reject_match(session_id, token, username):
    """Reject a match."""
    url = f"{MATCHMAKING_SERVICE_URL}/matchmaking/reject"
    headers = {"Authorization": f"Bearer {token}"}
    payload = {"session_id": session_id}

    print(f"❌ {username} rejecting match {session_id[:8]}...")
    response = requests.post(url, json=payload, headers=headers)

    if response.status_code != 200:
        print(f"❌ Failed to reject match for {username}: {response.status_code}")
        print(f"   Response: {response.text}")
        sys.exit(1)

    data = response.json()
    print(f"   Response: {data.get('message', 'N/A')}")
    return data


def main():
    print("=" * 60)
    print("🎯 Matchmaking Integration Test")
    print("=" * 60)
    print()

    # Step 1: Register two users
    print("Step 1: Registering users...")
    username1 = generate_random_username()
    username2 = generate_random_username()
    password = "password123"

    user1 = register_user(username1, password)
    user2 = register_user(username2, password)

    user1_id = user1["user_id"]
    user2_id = user2["user_id"]

    print()

    # Step 2: Login users and get tokens
    print("Step 2: Logging in users...")
    auth1 = login_user(username1, password)
    auth2 = login_user(username2, password)

    token1 = auth1["token"]
    token2 = auth2["token"]

    print()

    # Step 3: Join matchmaking queue
    print("Step 3: Joining matchmaking queue...")
    join_queue(token1, username1)
    join_queue(token2, username2)

    print()

    # Step 4: Wait for matchmaker to run (runs every 2 seconds)
    print("Step 4: Waiting for matchmaker to process queue...")
    wait_time = 4  # Wait 4 seconds to ensure at least one tick
    for i in range(wait_time, 0, -1):
        print(f"   ⏳ Waiting {i} seconds...")
        time.sleep(1)

    print()

    # Step 5: Check if both users were matched
    print("Step 5: Checking match results...")
    match1 = check_match(token1, username1)
    match2 = check_match(token2, username2)

    print()

    # Step 6: Verify results
    print("Step 6: Verifying match...")
    print(f"   User {user1_id} matched: {match1.get('matched', False)}")
    print(f"   User {user2_id} matched: {match2.get('matched', False)}")

    if not match1.get("matched"):
        print(f"❌ FAILED: User {user1_id} ({username1}) was not matched")
        sys.exit(1)

    if not match2.get("matched"):
        print(f"❌ FAILED: User {user2_id} ({username2}) was not matched")
        sys.exit(1)

    session1 = match1.get("session_id")
    session2 = match2.get("session_id")

    if session1 != session2:
        print(f"❌ FAILED: Users matched to different sessions")
        print(f"   User {user1_id} session: {session1}")
        print(f"   User {user2_id} session: {session2}")
        sys.exit(1)

    opponent1 = match1.get("opponent")
    opponent2 = match2.get("opponent")

    if opponent1 != username2:
        print(f"❌ FAILED: User {user1_id} opponent should be {username2}, got {opponent1}")
        sys.exit(1)

    if opponent2 != username1:
        print(f"❌ FAILED: User {user2_id} opponent should be {username1}, got {opponent2}")
        sys.exit(1)

    print()
    print("=" * 60)
    print("✅ SUCCESS! Both users were matched together")
    print("=" * 60)
    print(f"📊 Match Details:")
    print(f"   Session ID: {session1}")
    print(f"   Player 1: {username1} (MMR: {match1.get('your_mmr', 'N/A')})")
    print(f"   Player 2: {username2} (MMR: {match2.get('your_mmr', 'N/A')})")
    print("=" * 60)
    print()

    # Step 7: Both players accept the match
    print("Step 7: Testing match acceptance flow...")

    # Player 1 accepts first
    accept_result1 = accept_match(session1, token1, username1)
    if accept_result1.get("status") != "waiting":
        print(f"❌ FAILED: Expected status 'waiting' after first accept, got {accept_result1.get('status')}")
        sys.exit(1)

    if not accept_result1.get("player1_ready"):
        print(f"❌ FAILED: Player 1 should be ready after accepting")
        sys.exit(1)

    if accept_result1.get("player2_ready"):
        print(f"❌ FAILED: Player 2 should not be ready yet")
        sys.exit(1)

    print(f"   ✅ Player 1 ready, waiting for Player 2")

    # Small delay to simulate real-world timing
    time.sleep(1)

    # Player 2 accepts second
    accept_result2 = accept_match(session1, token2, username2)
    if accept_result2.get("status") != "in_progress":
        print(f"❌ FAILED: Expected status 'in_progress' after both accept, got {accept_result2.get('status')}")
        sys.exit(1)

    if not accept_result2.get("player1_ready") or not accept_result2.get("player2_ready"):
        print(f"❌ FAILED: Both players should be ready after accepting")
        sys.exit(1)

    print(f"   ✅ Both players ready, game started!")

    print()
    print("=" * 60)
    print("✅ FULL TEST PASSED!")
    print("=" * 60)
    print("   • Users registered successfully")
    print("   • Both joined matchmaking queue")
    print("   • Match found within expected time")
    print("   • Player 1 accepted (status: waiting)")
    print("   • Player 2 accepted (status: in_progress)")
    print("   • Game session started successfully")
    print("=" * 60)


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n⚠️  Test interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: services/matchmaking/integration_test.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"testing"
	"time"
)

const (
	authServiceURL        = "http://localhost:8000"
	matchmakingServiceURL = "http://localhost:8001"
)

// IntegrationTest tests the full matchmaking flow with real services
func TestMatchmakingIntegration(t *testing.T) {
	// Skip if not running integration tests
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	// Step 1: Register two users via auth service
	user1 := registerUser(t, "testplayer1", "password123")
	user2 := registerUser(t, "testplayer2", "password123")

	t.Logf("Created users: %d and %d", user1.UserID, user2.UserID)

	// Give a small delay for services to sync
	time.Sleep(100 * time.Millisecond)

	// Step 2: Both users join the matchmaking queue
	joinQueue(t, user1.UserID)
	joinQueue(t, user2.UserID)

	t.Logf("Both users joined the queue")

	// Step 3: Wait for matchmaker loop to run (runs every 2 seconds)
	// Add extra buffer time to ensure at least one tick happens
	time.Sleep(3 * time.Second)

	// Step 4: Check if both users have been matched
	match1 := checkMatch(t, user1.UserID)
	match2 := checkMatch(t, user2.UserID)

	// Step 5: Verify they were matched together
	if !match1.Matched {
		t.Fatalf("User 1 was not matched")
	}
	if !match2.Matched {
		t.Fatalf("User 2 was not matched")
	}

	if match1.SessionID != match2.SessionID {
		t.Fatalf("Users were matched to different sessions: %s vs %s", match1.SessionID, match2.SessionID)
	}

	if match1.Opponent != "testplayer2" {
		t.Errorf("User 1's opponent should be testplayer2, got %s", match1.Opponent)
	}
	if match2.Opponent != "testplayer1" {
		t.Errorf("User 2's opponent should be testplayer1, got %s", match2.Opponent)
	}

	t.Logf("✅ Successfully matched users into session: %s", match1.SessionID)
	t.Logf("   User 1 (%s) MMR: %d", "testplayer1", match1.YourMMR)
	t.Logf("   User 2 (%s) MMR: %d", "testplayer2", match2.YourMMR)
}

// registerUser registers a new user and returns their user ID
func registerUser(t *testing.T, username, password string) RegisterResponse {
	reqBody := map[string]string{
		"username": username,
		"password": password,
	}

	jsonData, _ := json.Marshal(reqBody)
	resp, err := http.Post(
		authServiceURL+"/auth/register",
		"application/json",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		t.Fatalf("Failed to register user %s: %v", username, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		t.Fatalf("Failed to register user %s: status %d, body: %s", username, resp.StatusCode, string(body))
	}

	var result RegisterResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		t.Fatalf("Failed to decode register response: %v", err)
	}

	return result
}

// joinQueue adds a user to the matchmaking queue
func joinQueue(t *testing.T, userID int64) {
	reqBody := map[string]int64{
		"user_id": userID,
	}

	jsonData, _ := json.Marshal(reqBody)
	resp, err := http.Post(
		matchmakingServiceURL+"/matchmaking/join",
		"application/json",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		t.Fatalf("Failed to join queue for user %d: %v", userID, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		t.Fatalf("Failed to join queue for user %d: status %d, body: %s", userID, resp.StatusCode, string(body))
	}

	t.Logf("User %d joined matchmaking queue", userID)
}

// checkMatch checks if a user has been matched
func checkMatch(t *testing.T, userID int64) MatchCheckResponse {
	resp, err := http.Get(
		fmt.Sprintf("%s/matchmaking/match?user_id=%d", matchmakingServiceURL, userID),
	)
	if err != nil {
		t.Fatalf("Failed to check match for user %d: %v", userID, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		t.Fatalf("Failed to check match for user %d: status %d, body: %s", userID, resp.StatusCode, string(body))
	}

	var result MatchCheckResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		t.Fatalf("Failed to decode match response: %v", err)
	}

	return result
}

// Response types
type RegisterResponse struct {
	UserID   int64  `json:"user_id"`
	Username string `json:"username"`
	Token    string `json:"token"`
}

type MatchCheckResponse struct {
	Matched   bool   `json:"matched"`
	SessionID string `json:"session_id"`
	Opponent  string `json:"opponent"`
	YourMMR   int    `json:"your_mmr"`
	TheirMMR  int    `json:"their_mmr"`
}
